python/cuspatial/cuspatial/core/_column/geocolumn.py

```diff
@@ -30,23 +31,102 @@ def __init__(
+        if from_read_polygon_shapefile:
+            """
+            A cudf.ListSeries needs four levels of nesting to represent
+            a polygon shapefile. The rings_offsets and polygons_offsets
+            have already been computed in the `read_polygon_shapefile`
+            function. In order to convert it into an arrow list<...>
+            we need a set of offsets buffers for each point tuple, and
+            an offsets buffer for the 1-offset multipolygons.
+
+            Coordinates: List of length 2 offsets: [0, 2, 4, ... n/2]
+            Rings: List of polygon ring offsets
+            Polygons: Offset into rings of each polygon
+            Multipolygons: List of length 1 offsets: No multipolygons
+
+            Finally, each set of offsets must have the length of the
+            array appended to the end, as Arrow offset lists are length
+            n + 1 but our original shapefile code offset lists are only
+            length n.
+            """
+            polygons_col = data[0].astype("int32")
+            rings_col = data[1].astype("int32")
+            coordinates = (
+                data[2].stack().astype("float64").reset_index(drop=True)
+            )
+            """
+            Store a fixed-size offsets buffer of even numbers:
+            0   0
+            1   2
+            2   4
+            ...
+            Up to the size of the original input.
+            """
+            coordinate_offsets = cudf.Series(
+                cp.arange(len(coordinates) + 1, step=2), dtype="int32"
+            )
+            rings_offsets = cudf.concat(
+                [
+                    cudf.Series(rings_col),
+                    cudf.Series([len(coordinate_offsets) - 1], dtype="int32"),
+                ]
+            ).reset_index(drop=True)
+            polygons_offsets = cudf.concat(
+                [
+                    cudf.Series(polygons_col),
+                    cudf.Series([len(polygons_col)], dtype="int32"),
+                ]
+            ).reset_index(drop=True)
+            coords = cudf.core.column.ListColumn(
+                size=len(coordinate_offsets) - 1,
+                dtype=cudf.ListDtype(coordinates.dtype),
+                children=(coordinate_offsets._column, coordinates._column),
+            )
+            rings = cudf.core.column.ListColumn(
+                size=len(rings_offsets) - 1,
+                dtype=cudf.ListDtype(coords.dtype),
+                children=(rings_offsets._column, coords),
+            )
+            polygons = cudf.core.column.ListColumn(
+                size=len(polygons_offsets) - 1,
+                dtype=cudf.ListDtype(rings.dtype),
+                children=(polygons_offsets._column, rings),
+            )
+            mpolygons = cudf.core.column.ListColumn(
+                size=len(polygons_offsets) - 1,
+                dtype=cudf.ListDtype(polygons.dtype),
+                children=(polygons_offsets._column, polygons),
+            )
```

Comment:
Hmmm, I think the multi-polygon offsets should be a sequence? e.g. `[0, 1, 2, 3, 4...]` so that the each of the multi-polygon contains only a single polygon.

```suggestion
            mpolygons = cudf.core.column.ListColumn(
                size=len(polygons_offsets) - 1,
                dtype=cudf.ListDtype(polygons.dtype),
                children=(cupy.arange(len(polygons) + 1), polygons),
            )
```

Reply:
Oh good point, I was writing that code based on my test data. :eyes: